cirq/neutral_atoms/neutral_atom_devices.py

```diff
@@ -0,0 +1,424 @@
+        """
+        Raises an error if the given operation is invalid on this device.
+
+        Args:
+            operation: the operation to validate
+
+        Raises:
+            ValueError: If the operation is not valid
+        """
+        if not isinstance(operation, (ops.GateOperation,
+                                      ops.ParallelGateOperation)):
+            raise ValueError('Unsupported operation: {!r}'.format(operation))
+
+        # The gate must be valid
+        self.validate_gate(operation.gate)
+
+        # All qubits the operation acts on must be on the device
+        for q in operation.qubits:
+            if q not in self.qubits:
+                raise ValueError('Qubit not on device: {!r}'.format(q))
+
+        if isinstance(operation.gate, (ops.MeasurementGate, ops.IdentityGate)):
+            return
+
+        # Verify that a controlled gate operation is valid
+        if isinstance(operation, ops.GateOperation):
+            if len(operation.qubits) > self._max_parallel_c:
+                raise ValueError("Too many qubits acted on in parallel by a "
+                                 "controlled gate operation")
+            if len(operation.qubits) > 1:
+                for p in operation.qubits:
+                    for q in operation.qubits:
+                        if (isinstance(p, GridQubit) and
+                                isinstance(q, GridQubit)):
+                            distance = sqrt((p.row - q.row)**2 +
+                                            (p.col - q.col)**2)
+                            if distance > self._control_radius:
+                                raise ValueError("Qubits {!r}, {!r} are too "
+                                                 "far away".format(p, q))
+            return
+
+        # Verify that a valid number of Z gates are applied in parallel
+        if isinstance(operation.gate, ops.ZPowGate):
+            if len(operation.qubits) > self._max_parallel_z:
+                raise ValueError("Too many Z gates in parallel")
+
+        # Verify that a valid number of XY gates are applied in parallel
+        if isinstance(operation.gate, (ops.XPowGate,
+                                       ops.YPowGate,
+                                       ops.PhasedXPowGate)):
+            if (len(operation.qubits) > self._max_parallel_xy and
+                    len(operation.qubits) != len(self.qubits)):
+                raise ValueError("Bad number of XY gates in parallel")
+
+    def validate_moment(self, moment: ops.Moment):
+        """
+        Raises an error if the given moment is invalid on this device
+
+        Args:
+            moment: The moment to validate
+
+        Raises:
+            ValueError: If the given moment is invalid
+        """
+        super().validate_moment(moment)
+
+        operations = moment.operations
+        num_parallel_z = 0
+        num_parallel_xy = 0
+        num_parallel_controlled_qubits = 0
```

Comment:
This variable tracks the number of qubits acted on by controlled gates. The length of the associated list is the number of controlled operations. So `num_parallel_controlled_qubits = sum( [len(l) for l in controlled_qubits_lists])`. Since this variable is used in multiple conditional checks, it made more sense to keep track of it alongside the list instead of re-calculating it using the list each time I needed to know it.

Reply:
I actually only need to know if the variable is greater than zero and one other number. It will be greater than zero if the length of the lists is, so I can just move the calculation of the variable to the check where it compares its value to `self._max_parallel_c`. This removes the advantage of storing the value in a variable so I've removed it.